// @payvo/sdk-ark/distribution/cjs/networks/ark.devnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shared_1 = require("./shared");
const network = {
    coin: "ARK",
    constants: {
        slip44: 1,
    },
    currency: {
        decimals: 8,
        symbol: "DѦ",
        ticker: "DARK",
    },
    explorer: shared_1.explorer,
    featureFlags: shared_1.featureFlags,
    governance: {
        delegateCount: 51,
        votesPerTransaction: 1,
        votesPerWallet: 1,
    },
    hosts: [
        {
            host: "https://ark-test.payvo.com/api",
            type: "full",
        },
        {
            host: "https://ark-test-musig.payvo.com",
            type: "musig",
        },
        {
            host: "https://dexplorer.ark.io",
            type: "explorer",
        },
    ],
    id: "ark.devnet",
    importMethods: shared_1.importMethods,
    knownWallets: "https://raw.githubusercontent.com/ArkEcosystem/common/master/devnet/known-wallets-extended.json",
    meta: {
        fastDelegateSync: true,
    },
    name: "Devnet",
    transactions: {
        ...shared_1.transactions,
        fees: {
            ticker: "DARK",
            type: "dynamic",
        },
        multiPaymentRecipients: 128,
    },
    type: "test",
};
exports.default = network;
