// @payvo/sdk-ark/distribution/cjs/networks/ark.mainnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shared_1 = require("./shared");
const network = {
    coin: "ARK",
    constants: {
        slip44: 111,
    },
    currency: {
        decimals: 8,
        symbol: "Ѧ",
        ticker: "ARK",
    },
    explorer: shared_1.explorer,
    featureFlags: shared_1.featureFlags,
    governance: {
        delegateCount: 51,
        votesPerTransaction: 1,
        votesPerWallet: 1,
    },
    hosts: [
        {
            host: "https://ark-live.payvo.com/api",
            type: "full",
        },
        {
            host: "https://ark-live-musig.payvo.com",
            type: "musig",
        },
        {
            host: "https://explorer.ark.io",
            type: "explorer",
        },
    ],
    id: "ark.mainnet",
    importMethods: shared_1.importMethods,
    knownWallets: "https://raw.githubusercontent.com/ArkEcosystem/common/master/mainnet/known-wallets-extended.json",
    meta: {
        fastDelegateSync: true,
    },
    name: "Mainnet",
    transactions: shared_1.transactions,
    type: "live",
};
exports.default = network;
