// @payvo/sdk-ark/distribution/cjs/networks/sxp.testnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shared_1 = require("./shared");
const network = {
    coin: "Solar",
    constants: {
        slip44: 1,
    },
    currency: {
        decimals: 8,
        symbol: "dSXP",
        ticker: "dSXP",
    },
    explorer: {
        block: "devnet/block/{0}",
        transaction: "devnet/transaction/{0}",
        wallet: "devnet/wallet/{0}",
    },
    featureFlags: {
        ...shared_1.featureFlags,
        Transaction: [
            "delegateRegistration",
            "delegateResignation",
            "estimateExpiration",
            "multiPayment",
            "secondSignature",
            "transfer",
            "vote",
        ],
    },
    governance: {
        delegateCount: 53,
        method: "split",
        votesPerTransaction: 1,
        votesPerWallet: 1,
    },
    hosts: [{
            host: "https://sxp.testnet.sh/api",
            type: "full",
        },
        {
            host: "https://explorer.solar.network/",
            type: "explorer",
        },
    ],
    id: "sxp.testnet",
    importMethods: shared_1.importMethods,
    meta: {
        fastDelegateSync: true,
    },
    name: "Testnet",
    transactions: {
        ...shared_1.transactions,
        fees: {
            ticker: "dSXP",
            type: "dynamic",
        },
        multiPaymentRecipients: 128,
    },
    type: "test",
};
exports.default = network;
