// @payvo/sdk-ark/distribution/cjs/networks/xqr.mainnet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const shared_1 = require("./shared");
const network = {
    coin: "Qredit",
    constants: {
        slip44: 111,
    },
    currency: {
        decimals: 8,
        symbol: "Q",
        ticker: "XQR",
    },
    explorer: {
        block: "qredit/block/{0}",
        transaction: "qredit/transaction/{0}",
        wallet: "qredit/wallet/{0}",
    },
    featureFlags: {
        ...shared_1.featureFlags,
        Transaction: [
            "delegateRegistration",
            "delegateResignation",
            "estimateExpiration",
            "multiPayment",
            "secondSignature",
            "transfer",
            "vote",
        ],
    },
    governance: {
        delegateCount: 51,
        method: "split",
        votesPerTransaction: 1,
        votesPerWallet: 1,
    },
    hosts: [{
            host: "https://qredit.cloud/api",
            type: "full",
        },
        {
            host: "https://explorer.sh/",
            type: "explorer",
        },
    ],
    id: "xqr.mainnet",
    importMethods: shared_1.importMethods,
    knownWallets: "https://raw.githubusercontent.com/qredit/common/master/mainnet/known-wallets-extended.json",
    meta: {
        fastDelegateSync: true,
    },
    name: "Mainnet",
    transactions: {
        ...shared_1.transactions,
        fees: {
            ticker: "XQR",
            type: "dynamic",
        },
        multiPaymentRecipients: 128,
    },
    type: "live",
};
exports.default = network;
