// @payvo/sdk-ark/distribution/esm/networks/sxp.mainnet.js
import { featureFlags, importMethods, transactions } from "./shared";
const network = {
    coin: "Solar",
    constants: {
        slip44: 111,
    },
    currency: {
        decimals: 8,
        symbol: "SXP",
        ticker: "SXP",
    },
    explorer: {
        block: "mainnet/block/{0}",
        transaction: "mainnet/transactions/{0}",
        wallet: "mainnet/wallet/{0}",
    },
    featureFlags: {
        ...featureFlags,
        Transaction: [
            "delegateRegistration",
            "delegateResignation",
            "estimateExpiration",
            "multiPayment",
            "secondSignature",
            "transfer",
            "vote",
        ],
    },
    governance: {
        delegateCount: 53,
        method: "split",
        votesPerTransaction: 1,
        votesPerWallet: 1,
    },
    hosts: [{
            host: "https://sxp.mainnet.sh/api",
            type: "full",
        },
        {
            host: "https://explorer.solar.network",
            type: "explorer",
        },
    ],
    id: "sxp.mainnet",
    importMethods,
    knownWallets: "https://raw.githubusercontent.com/solar-network/common/master/mainnet/known-wallets-extended.json",
    meta: {
        fastDelegateSync: true,
    },
    name: "Mainnet",
    transactions: {
        ...transactions,
        fees: {
            ticker: "SXP",
            type: "dynamic",
        },
        multiPaymentRecipients: 128,
    },
    type: "live",
};
export default network;
